staging/kos/README.md

```diff
@@ -188,6 +196,119 @@ remote NetworkAttachment node. The flows installed in that bridge will
+to check if an invariant is violated. But different API objects are not
+necessarily validated sequentially by the extension API server. If a
+subnet *Y* in *conflict* with *X* (i.e, *X* and *Y* mutual existences
+lead to a violation of one invariant) is validated in parallel with *X*,
+they will both be created. This case must be handled, but it is a rare
+occurrence, so we chose to keep the validation by the API server
+nevertheless.
+
+To solve the problem also in the cases where two conflicting subnets
+are created, we replaced the hard-to-enforce constraint that two subnets
+violating the invariants cannot exist with the easier-to-enforce constraint
+that if two subnets violating the invariants exist, subnet consumers use
+only one and ignore the other. To achieve this, each subnet has a `bool`
+field called `SubnetStatus.Validated`. Upon creation it is set to false by
+the extension API server, which also sets it to false after updates to
+`SubnetSpec.IPv4` (the CIDR) and `SubnetSpec.VNI`. Consumers of subnets
+(currently only the [IPAM controller](#the-ipam-controller)) can use a
+subnet only when `SubnetStatus.Validated` is `true`.
+
+The subnets validator task is setting `SubnetStatus.Validated` as appropriate
+for every subnet. It uses an Informer to remain apprised of all the subnets
+and follows the usual Kubernetes controller design pattern based on a rate
+limiting queue and worker goroutines. When a worker processes an existing
+subnet *X*, it does a live list of all the existing subnets against the
+extension API server, and compares *X* against the results of the list. If no
+violation of the invariants is found, *X*'s `SubnetStatus.Validated` is set
+to true. The validator also keeps a *conflicts cache* which associates each
+processed subnet *Y* to its last seen *VNI* and *CIDR* and a list with the
+namespaced names of the subnets whose validation failed because of a conflict
+with *Y* when *Y* had the last seen *VNI* and *CIDR*. If, while validating
+*X*, a conflict with *Y* is found, *Y*'s *VNI* and *CIDR* are compared to those
+stored in its *conflicts cache* entry. If they match, *X*'s status is updated
+by adding information on the conflict, and *X*'s namespaced name is added to
+the conflicts cache entry for *Y*. If they do not match, no further action is
+taken and *X* is processed again after a delay. When *Y* is deleted or its
+*VNI* and *CIDR* are updated, its *confilcts cache* entry is updated: the new
+*VNI* and *CIDR* override the old ones (in case of an update) and the list of
+conflicting subnets namespaced names is reset. All the namespaced names in the
+old list are enqueued so that their subnets can be re-validated: they did not
+pass validation because of a conflict with *Y*, but *Y* no longer exists or has
+changed (potentially making the conflict disappear), so they might be
+successfully validated now. If *X* was added unconditionally to *Y*'s conflicts
```

Comment:
It should be explicitly stated that the "if" here is not about a possible executing in the current code but rather is about conceivable alternate code.  I favor a slightly more radical re-write, simply explaining the virtue of the existing code directly rather than discussing how alternate code could fail.  Perhaps something like the following.

Making the addition of *X* to the conflicts cache of *Y* fail on mistach with the VNI or CIDR in the conflict cache of *Y* correctly handles the possible concurrent execution where one goroutine processing *X* finds a conflict with *Y*, then a different goroutine processes an update to the VNI or CIDR of *Y*, then the first goroutine gets around to trying to add *X* to the conflict cache of *Y*; adding *X* to the conflict cache of *Y* in this situation would falsely indicate that *X* does not need to be reconsidered until a _further_ change is made to *Y*.

Reply:
I agree both with the problem and your proposed solution. Replaced the old phrasing with your suggestion in 75fc302.